[WebXR][OpenXR] Unreliable click events using hand interaction profile

svillar opened the pull request:
#### 1d859edbec260d5dc6b10cea38256e92448239ec
<pre>
[WebXR][OpenXR] Unreliable click events using hand interaction profile
<a href="https://bugs.webkit.org/show_bug.cgi?id=298091">https://bugs.webkit.org/show_bug.cgi?id=298091</a>

Reviewed by Adrian Perez de Castro.

The pinch gesture is measured by the pinch_ext/value path which
defines a float in the [0, 1] range, with 1 meaning both fingers
touching. The OpenXR spec does not define a press or touch paths
for pinches so clients must decide when the value means a pinch
(or a click).

Values returned by runtimes vary a lot so we cannot trust them.
It&apos;s better so set our own threshold to consider a pinch gesture
as taking action. Experience tells us that 0.9 is a very good one.

There is no need to apply those thresholds for controllers, because
runtimes provide touch and press events for physical elements like
triggers or buttons. That&apos;s why we apply this threshold only when
the hand interaction profile is used.

* Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
(WebKit::OpenXRInputSource::updateInteractionProfile):
(WebKit::OpenXRInputSource::collectButton const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:

Canonical link: <a href="https://commits.webkit.org/299321@main">https://commits.webkit.org/299321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd282845c7268448bd2bfe5f51d0a5faa62ae868

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118604 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28936 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124781 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70663 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38981 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46867 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90017 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59563 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121557 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31049 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106335 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70521 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30112 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68438 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100487 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24637 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127842 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45511 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34338 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98650 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45875 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102556 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98434 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25028 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43878 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21873 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42009 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45381 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51059 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44844 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48191 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46531 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dd282845c7268448bd2bfe5f51d0a5faa62ae868. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50060

svillar committed:
[WebXR][OpenXR] Unreliable click events using hand interaction profile
https://bugs.webkit.org/show_bug.cgi?id=298091

Reviewed by Adrian Perez de Castro.

The pinch gesture is measured by the pinch_ext/value path which
defines a float in the [0, 1] range, with 1 meaning both fingers
touching. The OpenXR spec does not define a press or touch paths
for pinches so clients must decide when the value means a pinch
(or a click).

Values returned by runtimes vary a lot so we cannot trust them.
It's better so set our own threshold to consider a pinch gesture
as taking action. Experience tells us that 0.9 is a very good one.

There is no need to apply those thresholds for controllers, because
runtimes provide touch and press events for physical elements like
triggers or buttons. That's why we apply this threshold only when
the hand interaction profile is used.

* Source/WebKit/UIProcess/XR/openxr/OpenXRInputMappings.h:
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.cpp:
(WebKit::OpenXRInputSource::updateInteractionProfile):
(WebKit::OpenXRInputSource::collectButton const):
* Source/WebKit/UIProcess/XR/openxr/OpenXRInputSource.h:

Canonical link: https://commits.webkit.org/299321@main

aperezdc commented:
Backported into `webkitglib/2.50` as commit 5519f11b83fd6524cd39779225d332693f58f6e6
